ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java

```diff
@@ -182,22 +180,28 @@ public String toString() {
 
   @Override
   public void run() {
+    CompletableFuture<Void> applyLogFutures = CompletableFuture.completedFuture(null);
     for(; state != State.STOP; ) {
       try {
-        waitForCommit();
+        waitForCommit(applyLogFutures);
 
         if (state == State.RELOAD) {
           reload();
         }
 
-        final MemoizedSupplier<List<CompletableFuture<Message>>> futures = applyLog();
-        checkAndTakeSnapshot(futures);
+        applyLogFutures = applyLog(applyLogFutures);
+        checkAndTakeSnapshot(applyLogFutures);
 
         if (shouldStop()) {
-          checkAndTakeSnapshot(futures);
+          applyLogFutures.get();
           stop();
         }
       } catch (Throwable t) {
+        try {
+          applyLogFutures.get();
+        } catch (InterruptedException | ExecutionException e) {
+          LOG.info("{}: interrupted while waiting for apply transactions", this, t);
+        }
```

Comment:
@swamirishi , Why calling ` applyLogFutures.get()` when there is already an exception/error?   It could cause problems such as:
1. The future may never complete.
2. The thread may already be interrupted.
3. In may throw another runtime exception/error (such as OOM)

The test failures could be caused by this change.

Reply:
I have updated the catch block to not wait when the exception is ExecutionException & InterruptedException. We should wait for apply transactions in all other cases

Reply:
> @swamirishi , Why calling ` applyLogFutures.get()` when there is already an exception/error? It could cause problems such as:
> 
> 1. The future may never complete.
It is guaranteed it will always comple right?
> 2. The thread may already be interrupted.
> 3. In may throw another runtime exception/error (such as OOM)
It won't be thrown by the future right? If the future throws exception we anyways handle it and mark the future as completed.
https://github.com/apache/ratis/pull/1218/files#:~:text=applyLogFutures%20%3D%20applyLogFutures,)%20%2D%3E%20null)%3B

Reply:
> It is guaranteed it will always complete right?

@swamirishi , Who will guarantee that?  Ozone?  Any doc to support this statement?  It seems impossible in an erroneous condition.

Reply:
> It won't be thrown by the future right? ...

It definitely could throw OOM.